### PR TITLE
mgr/volumes: Fix subvolume creation in FIPS enabled system.

### PR DESCRIPTION
**Pull request by kotreshhr**
The md5 checksum is used in the construction of legacy
subvolume config filename. It's not used for security reason.
Hence marking the 'usedforsecurity' flag to false to
make it FIPs compliant.

The usage of md5 was always in there. The commit 373a04cf734
made it to get exercised in 'open_subvol' which is pre-requisite
for all the subvolume operations and hence subvolume
creation has failed.

Fixes: https://tracker.ceph.com/issues/56727
Signed-off-by: Kotresh HR <khiremat@redhat.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Review by ajarr — APPROVED** [review]

**Comment by ajarr** [comment]
@kotreshhr pointed out the make check failure. Thanks, @kotreshhr 

The failure
```
mypy run-test-pre: PYTHONHASHSEED='800757976'
mypy run-test: commands[0] | mypy --config-file=../../mypy.ini -m alerts -m balancer -m cephadm -m crash -m dashboard -m devicehealth -m diskprediction_local -m hello -m influx -m iostat -m localpool -m mds_autoscaler -m mgr_module -m mgr_util -m mirroring -m nfs -m orchestrator -m pg_autoscaler -m progress -m prometheus -m rbd_support -m rook -m snap_schedule -m selftest -m stats -m status -m telegraf -m telemetry -m test_orchestrator -m volumes -m zabbix
volumes/fs/operations/versions/__init__.py:7: note: In module imported here,
volumes/fs/operations/subvolume.py:5: note: ... from here,
volumes/fs/volume.py:16: note: ... from here,
volumes/module.py:9: note: ... from here,
volumes/__init__.py:2: note: ... from here:
volumes/fs/operations/versions/subvolume_base.py: note: In member "legacy_config_path" of class "SubvolumeBase":
volumes/fs/operations/versions/subvolume_base.py:79: error: Unexpected keyword argument "usedforsecurity" for "new"
.tox/mypy/lib/python3.8/site-packages/mypy/typeshed/stdlib/hashlib.pyi:29: note: "new" defined here
Found 1 error in 1 file (checked 31 source files)
```

**Comment by mchangir** [comment]
Python docs say ...
Changed in version 3.9: All hashlib constructors take a keyword-only argument _usedforsecurity_ with default value `True`.

and the error message below in https://github.com/ceph/ceph/pull/47300#issuecomment-1197408564
`.tox/mypy/lib/python3.8/site-packages/mypy/typeshed/stdlib/hashlib.pyi:29`: note: "new" defined here

points to version 3.8

**Comment by kotreshhr** [comment]
> Python docs say ... Changed in version 3.9: All hashlib constructors take a keyword-only argument _usedforsecurity_ with default value `True`.
> 
> and the error message below in [#47300 (comment)](https://github.com/ceph/ceph/pull/47300#issuecomment-1197408564) `.tox/mypy/lib/python3.8/site-packages/mypy/typeshed/stdlib/hashlib.pyi:29`: note: "new" defined here
> 
> points to version 3.8

@mchangir, I guess, you are right. And also there is a dependency on underlying 'open_ssl' library being used as well. On my ubuntu system with python 3.8, `hashlib.new('md5', usedforsecurity=False)` does take `usedforsecuirty` flag but I think does get ignored while calling open_ssl's md5()

```
# python3
Python 3.8.10 (default, Jun 22 2022, 20:18:18) 
[GCC 9.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import hashlib

>>> try:
...     m = hashlib.new('md5', usedforsecurity=False)
...     print("supports usedforsecurity flag")
... except TypeError:
...     m = hashlib.new('md5')
...     print("Doesn't support usedforsecuirty")
... 
supports usedforsecurity flag
```

But calling directly the open_ssl's md5() with `usedforsecurity` flag fails.

```
>>> m = hashlib.md5(b'vol', usedforsecurity=False)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: openssl_md5() takes at most 1 argument (2 given)
>>> 
```

Since we can't replace md5 with fips supported hash as it breaks backward compatibility/upgrade path, we have to rely on `usedforsecurity` flag. I think best approach would be it use the `usedforsecurity` flag on the systems it supports and fall back to earlier code on others as below.

```
>>> try:
...     m = hashlib.md5(b'vol', usedforsecurity=False)
... except TypeError:
...     m = hashlib.md5(b'vol')
... 

```


At least on RHEL, CentOS, Fedora, this flag seems to be adopted early [1][2] and should work. This means, mgr/volumes is FIPS compliant on systems where 'usedforsecurity' flag is supported.

 @ajarr @gregsfortytwo @batrick 
Let me know your thoughts.

[1] https://bugzilla.redhat.com/show_bug.cgi?id=1744670
[2] https://code.djangoproject.com/ticket/28401

**Comment by ajarr** [comment]
So if usedforsecurity argument is not supported  in a FIPS enabled system, then there'd be ValueError trace back in the mgr logs, "ValueError: [digital envelope routines: EVP_DigestInit_ex] disabled for FIPS" as mentioned in the tracker ticket. What does the end user who issued the `subvolume create` command see?  Should we handle the ValueError?

**Comment by kotreshhr** [comment]
> So if usedforsecurity argument is not supported in a FIPS enabled system, then there'd be ValueError trace back in the mgr logs, "ValueError: [digital envelope routines: EVP_DigestInit_ex] disabled for FIPS" as mentioned in the tracker ticket. What does the end user who issued the `subvolume create` command see? Should we handle the ValueError?

No, we get a `TypeError` if `usedforsecurity` flag is not supported on both fips enabled and non fips system. 
The `ValueError` is got when `md5' is used without `usedforsecurity=True` flag on FIPs enabled system.

**Comment by ajarr** [comment]
> > So if usedforsecurity argument is not supported in a FIPS enabled system, then there'd be ValueError trace back in the mgr logs, "ValueError: [digital envelope routines: EVP_DigestInit_ex] disabled for FIPS" as mentioned in the tracker ticket. What does the end user who issued the `subvolume create` command see? Should we handle the ValueError?
> 
> No, we get a `TypeError` if `usedforsecurity` flag is not supported on both fips enabled and non fips system. The `ValueError` is got when `md5` is used without `usedforsecurity=False` flag on FIPs enabled system.

I meant with your current commit, if the `usedforsecurity` flag is not supported  we revert to the old call. And if the old call fails with "ValueError: [digital envelope routines: EVP_DigestInit_ex] disabled for FIPS",  we don't handle that. Maybe we should?

**Comment by batrick** [comment]
> > > So if usedforsecurity argument is not supported in a FIPS enabled system, then there'd be ValueError trace back in the mgr logs, "ValueError: [digital envelope routines: EVP_DigestInit_ex] disabled for FIPS" as mentioned in the tracker ticket. What does the end user who issued the `subvolume create` command see? Should we handle the ValueError?
> > 
> > 
> > No, we get a `TypeError` if `usedforsecurity` flag is not supported on both fips enabled and non fips system. The `ValueError` is got when `md5` is used without `usedforsecurity=False` flag on FIPs enabled system.
> 
> I meant with your current commit, if the `usedforsecurity` flag is not supported we revert to the old call. And if the old call fails with "ValueError: [digital envelope routines: EVP_DigestInit_ex] disabled for FIPS", we don't handle that. Maybe we should?

At the very least we should log a meaningful error, yes.

**Review by batrick — APPROVED** [review]
LGTM

**Comment by kotreshhr** [comment]
> > > > So if usedforsecurity argument is not supported in a FIPS enabled system, then there'd be ValueError trace back in the mgr logs, "ValueError: [digital envelope routines: EVP_DigestInit_ex] disabled for FIPS" as mentioned in the tracker ticket. What does the end user who issued the `subvolume create` command see? Should we handle the ValueError?
> > > 
> > > 
> > > No, we get a `TypeError` if `usedforsecurity` flag is not supported on both fips enabled and non fips system. The `ValueError` is got when `md5` is used without `usedforsecurity=False` flag on FIPs enabled system.
> > 
> > 
> > I meant with your current commit, if the `usedforsecurity` flag is not supported we revert to the old call. And if the old call fails with "ValueError: [digital envelope routines: EVP_DigestInit_ex] disabled for FIPS", we don't handle that. Maybe we should?
> 
> At the very least we should log a meaningful error, yes.

Done

File: src/pybind/mgr/volumes/fs/operations/versions/subvolume_base.py [hunk 1]
```diff
@@ -76,9 +76,15 @@ def legacy_dir(self):
...
     @property
     def legacy_config_path(self):
-        m = md5()
-        m.update(self.base_path)
-        meta_config = "{0}.meta".format(m.digest().hex())
+        try:
+            m = hashlib.md5(self.base_path)
+        except ValueError:
+            try:
+                m = hashlib.md5(self.base_path, usedforsecurity=False) # type: ignore
```
Comment (kotreshhr) [inline src/pybind/mgr/volumes/fs/operations/versions/subvolume_base.py:83]: @ajarr @batrick Ignored mypy error and switched the order. Assuming non fips systems are comparatively more. PTAL

File: src/pybind/mgr/volumes/fs/operations/versions/subvolume_base.py [hunk 2]
```diff
@@ -76,9 +76,15 @@ def legacy_dir(self):
...
+            except TypeError:
+                raise VolumeException(-errno.EINVAL, "usedforsecurity flag is not supported on this system and FIPs enabled.")
```
Comment (ajarr) [inline src/pybind/mgr/volumes/fs/operations/versions/subvolume_base.py:85]: Maybe saying, "require python's hashlib library to support usedforsecurity flag in FIPS enabled systems", is more useful?

**Review by ajarr — APPROVED** [review]

**Comment by kotreshhr** [comment]
jenkins test make check

**Comment by kotreshhr** [comment]
jenkins test api

**Comment by kotreshhr** [comment]
jenkins retest this please

**Comment by kotreshhr** [comment]
@ajarr Triggered `fs:volumes` teuthology run.  

http://pulpito.front.sepia.ceph.com/khiremat-2022-07-29_17:04:35-fs:volumes-wip-khiremat-46599-fip-md5-mgr-volumes-distro-default-smithi/

**Comment by ajarr** [comment]
> @ajarr Triggered `fs:volumes` teuthology run.
> 
> http://pulpito.front.sepia.ceph.com/khiremat-2022-07-29_17:04:35-fs:volumes-wip-khiremat-46599-fip-md5-mgr-volumes-distro-default-smithi/

The manila fix related test runs failed,
```2022-07-29T17:36:30.386 INFO:tasks.cephfs_test_runner:======================================================================
2022-07-29T17:36:30.386 INFO:tasks.cephfs_test_runner:ERROR: test_binary_metafile_on_legacy_to_v1_upgrade (tasks.cephfs.test_volumes.TestMisc)
2022-07-29T17:36:30.387 INFO:tasks.cephfs_test_runner:----------------------------------------------------------------------
2022-07-29T17:36:30.387 INFO:tasks.cephfs_test_runner:Traceback (most recent call last):
2022-07-29T17:36:30.387 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_ceph-c_118af2d42d4f09fea80118ebcbec36a4c7315506/qa/tasks/cephfs/test_volumes.py", line 7472, in test_binary_metafile_on_legacy_to_v1_upgrade
2022-07-29T17:36:30.388 INFO:tasks.cephfs_test_runner:    self.mount_a.run_shell(['mkdir', '-p', createpath], sudo=True)
2022-07-29T17:36:30.389 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_ceph-c_118af2d42d4f09fea80118ebcbec36a4c7315506/qa/tasks/cephfs/mount.py", line 717, in run_shell
2022-07-29T17:36:30.389 INFO:tasks.cephfs_test_runner:    omit_sudo=omit_sudo, **kwargs)
2022-07-29T17:36:30.390 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_git_teuthology_dd7eb9d989259767783fb5ee9a761cd0732b84e6/teuthology/orchestra/remote.py", line 510, in run
2022-07-29T17:36:30.391 INFO:tasks.cephfs_test_runner:    r = self._runner(client=self.ssh, name=self.shortname, **kwargs)
2022-07-29T17:36:30.391 INFO:tasks.cephfs_test_runner:TypeError: run() got an unexpected keyword argument 'sudo'
2022-07-29T17:36:30.392 INFO:tasks.cephfs_test_runner:
```
I think we need the commit https://github.com/ceph/ceph/pull/47236/commits/1f6860a25ebf4512147539074e77417bd6089ad2 in this PR as well

**Comment by ajarr** [comment]
Running fs:volumes suite on the ceph branch "wip-khiremat-46599-fip-md5-mgr-volumes" with qa-suite pointing to https://github.com/ajarr/ceph/commits/fips-fix, which has your qa-suite fix
```
$ export CEPH_QA_MAIL="rraja@redhat.com"; MACHINE_NAME=smithi; CEPH_BRANCH=wip-khiremat-46599-fip-md5-mgr-volumes
$  teuthology-suite -v --ceph-repo https://github.com/ceph/ceph-ci.git --suite-repo https://github.com/ajarr/ceph  -c $CEPH_BRANCH -m $MACHINE_NAME -s fs:volumes -p 75 -e $CEPH_QA_MAIL --suite-branch fips-fix
```
http://pulpito.front.sepia.ceph.com/rraja-2022-07-29_22:43:34-fs:volumes-wip-khiremat-46599-fip-md5-mgr-volumes-distro-default-smithi/

Let's see if  tasks.cephfs.test_volumes.TestMisc passes now

**Comment by kotreshhr** [comment]
> Running fs:volumes suite on the ceph branch "wip-khiremat-46599-fip-md5-mgr-volumes" with qa-suite pointing to https://github.com/ajarr/ceph/commits/fips-fix, which has your qa-suite fix
> 
> ```
> $ export CEPH_QA_MAIL="rraja@redhat.com"; MACHINE_NAME=smithi; CEPH_BRANCH=wip-khiremat-46599-fip-md5-mgr-volumes
> $  teuthology-suite -v --ceph-repo https://github.com/ceph/ceph-ci.git --suite-repo https://github.com/ajarr/ceph  -c $CEPH_BRANCH -m $MACHINE_NAME -s fs:volumes -p 75 -e $CEPH_QA_MAIL --suite-branch fips-fix
> ```
> 
> http://pulpito.front.sepia.ceph.com/rraja-2022-07-29_22:43:34-fs:volumes-wip-khiremat-46599-fip-md5-mgr-volumes-distro-default-smithi/
> 
> Let's see if tasks.cephfs.test_volumes.TestMisc passes now

Unfortunately, the qa source code has diverted between release branches especially the use of `sudo` argument! The one that works in `main` might not work in `octopus`.

**Comment by ajarr** [comment]
I'm testing your latest qa fix with your existing builds,
```
$ teuthology-suite -v --ceph-repo https://github.com/ceph/ceph-ci.git --suite-repo https://github.com/kotreshhr/ceph  -c $CEPH_BRANCH -m $MACHINE_NAME -s fs:volumes -p 75 -e $CEPH_QA_MAIL --suite-branch fips-md5-issue
```

http://pulpito.front.sepia.ceph.com/rraja-2022-07-30_01:34:24-fs:volumes-wip-khiremat-46599-fip-md5-mgr-volumes-distro-default-smithi/

**Comment by ajarr** [comment]
> I'm testing your latest qa fix with your existing builds,
> 
> ```
> $ teuthology-suite -v --ceph-repo https://github.com/ceph/ceph-ci.git --suite-repo https://github.com/kotreshhr/ceph  -c $CEPH_BRANCH -m $MACHINE_NAME -s fs:volumes -p 75 -e $CEPH_QA_MAIL --suite-branch fips-md5-issue
> ```
> 
> http://pulpito.front.sepia.ceph.com/rraja-2022-07-30_01:34:24-fs:volumes-wip-khiremat-46599-fip-md5-mgr-volumes-distro-default-smithi/

TestsMisc pased

File: src/pybind/mgr/volumes/fs/operations/versions/subvolume_base.py (hunk 1 above)
Comment (ajarr) [reply src/pybind/mgr/volumes/fs/operations/versions/subvolume_base.py:83]: got it.

**Pull request merged**